### PR TITLE
Add view support for select queries

### LINKED ISSUE #442: View support on distributed queries
Currently, views on distributed queries do not work.

We are able to create a view on a distributed query;

``` sql
CREATE VIEW book_reviews AS
  (SELECT * FROM customer_reviews WHERE product_group = 'Book' LIMIT 1000);
```

But if we try to use the view, Citus errors out;

``` sql
SELECT COUNT(1) FROM book_reviews;

ERROR:  cannot plan queries that include both regular and partitioned relations
```

### PR DESCRIPTION
**Pull request by mtuncer**
Enables use views within distributed queries.
User can create and use a view on distributed tables/queries
as he/she would use with regular queries.

After this change router queries will have full support for views,
insert into select queries will support reading from views, not
writing into. Outer joins would have a limited support, and would
error out at certain cases such as when a view is in the inner side
of the outer join.

Although PostgreSQL supports writing into views under certain circumstances.
We disallowed that for distributed views. Also copying into a view is not supported.

There is no change in behavior for materialized views. They are only supported for router queries, and not supported for real-time or task tracker queries.

Fixes #442

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -970,6 +1016,23 @@ ErrorIfModifyQueryNotSupported(Query *queryTree)
 		   commandType == CMD_DELETE);
 
 	/*
+	 * Result relation is expected to be the first range table entry.
+	 * However, when it is a view, a new RTE is added. Original RTE
+	 * stays at the first place. This type of queries are not supported.
+	 */
+	if (queryTree->resultRelation != 1)
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I really don't want to hardcode that assumption. Which turns out not to actually be true in general...

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 2]
```diff
@@ -1560,9 +1625,11 @@ ExtractRangeTableRelationWalker(Node *node, List **rangeTableRelationList)
...
 	{
 		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
-		if (rangeTableEntry->rtekind == RTE_RELATION)
+		if (rangeTableEntry->rtekind == RTE_RELATION &&
+			rangeTableEntry->relkind == RELKIND_RELATION)
 		{
-			(*rangeTableRelationList) = lappend(*rangeTableRelationList, rangeTableEntry);
+			(*rangeTableRelationList) = lappend(*rangeTableRelationList,
+												rangeTableEntry);
 		}
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: Isn't this a completely unrelated (and unnecessary?) change?

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 3]
```diff
@@ -299,7 +301,7 @@ MultiPlanTree(Query *queryTree)
 		 * elements.
 		 */
 		joinClauseList = JoinClauseList(whereClauseList);
-		tableEntryList = TableEntryList(rangeTableList);
+		tableEntryList = UsedTableEntryList(queryTree);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: There's other places that use TableEntryList - why don't they need to be adjusted?

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 4]
```diff
@@ -554,6 +574,12 @@ ErrorIfSubqueryNotSupported(Query *subqueryTree)
 		errorDetail = "Subqueries with offset are not supported yet";
 	}
 
+	if (QueryReferencesView(subqueryTree))
+	{
+		preconditionsSatisfied = false;
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: Aren't there classes of this that we actually would support? I'm not saying we have to, but I think if so this should carry a comment saying that we just don't bother to figure out which are supported.

File: src/test/regress/input/multi_view.source [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+	l_shipmode char(10) not null,
+	l_comment varchar(44) not null );
+
+SET citus.shard_count to 4;
+SET citus.shard_replication_factor to 1;
+
+SELECT create_distributed_table('lineitem_hash', 'l_orderkey');
+
+\copy lineitem_hash FROM '@abs_srcdir@/data/lineitem.1.data' with delimiter '|'
+\copy lineitem_hash FROM '@abs_srcdir@/data/lineitem.2.data' with delimiter '|'
```
Comment (anarazel) [inline src/test/regress/input/multi_view.source]: Hm. Can't we move this to one central file? We now create, load, dump this in multiple places. And all of that requires us to use .source files, whic hare much more painful to maintain.  So please move lineitem_hash etc from all tests into a multi_create_table.sql/multi_load_data.source.

File: src/test/regress/input/multi_view.source [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	GROUP BY l_suppkey ORDER BY 2 DESC, 1 LIMIT 5;
+
+-- when a view is replaced by actual query it still fails
+SELECT l_suppkey, count(*) FROM
+	(SELECT l_suppkey, l_shipdate, count(*)
+		FROM (SELECT * FROM lineitem_hash WHERE l_shipmode = 'AIR') asi
+		GROUP BY l_suppkey, l_shipdate) supps
+	GROUP BY l_suppkey ORDER BY 2 DESC, 1 LIMIT 5;
+
+SET citus.task_executor_type to DEFAULT;
```
Comment (anarazel) [inline src/test/regress/input/multi_view.source]: Can we please add a test that includes a GROUP BY? I suspect those are actually the most frequent ones in practice, where the view "returns" some highly aggregated data, and you just perform some selection ontop of that.

Comment (anarazel) [inline src/test/regress/input/multi_view.source]: Can we please add a test that includes a GROUP BY? I suspect those are actually the most frequent ones in practice, where the view "returns" some highly aggregated data, and you just perform some selection ontop of that.

File: src/backend/distributed/planner/multi_logical_planner.c (hunk 3 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_planner.c]: They were being used for a different purpose. TableEntryList is called only from RelationIdList which is interested in all relation Ids used inside a query tree. This change is however deals with creating a join node, and any range table entry down below the subquery is irrelevant. 

I agree with you on not reporting unused range table entries in TableEntryList function too. I will check this again.

File: src/backend/distributed/planner/multi_logical_planner.c (hunk 4 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_planner.c]: There are possible cases where it would work, my tests resulted in non-working cases to be norm. 

👍  on a comment. will do that.

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 7]
```diff
@@ -1560,9 +1625,11 @@ ExtractRangeTableRelationWalker(Node *node, List **rangeTableRelationList)
 	foreach(rangeTableCell, rangeTableList)
 	{
 		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
-		if (rangeTableEntry->rtekind == RTE_RELATION)
+		if (rangeTableEntry->rtekind == RTE_RELATION &&
+			rangeTableEntry->relkind == RELKIND_RELATION)
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_planner.c]: right. correcting it.

File: src/backend/distributed/planner/multi_logical_planner.c (hunk 2 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_planner.c]: reverted

File: src/backend/distributed/planner/multi_router_planner.c [hunk 8]
```diff
@@ -628,6 +628,19 @@ ErrorIfInsertSelectQueryNotSupported(Query *queryTree, RangeTblEntry *insertRte,
 	Query *subquery = NULL;
 	Oid selectPartitionColumnTableId = InvalidOid;
 
+	if (queryTree->resultRelation != 1)
+	{
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: due to two observation,
resultRelation is expected to be 1, for insert queries. If it is not means postgresql intervened and added a new RTE to be used for insert. 
If a view is a target for insert, it first replaced by another view, then an actual relation rte.

I could not find a safer way of checking this. Open to suggestions.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 1 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: Any suggestions ?

File: src/test/regress/input/multi_view.source (hunk 5 above)
Comment (mtuncer) [inline src/test/regress/input/multi_view.source]: I did, but multi_drop_extension removed them.

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 9]
```diff
@@ -514,6 +516,24 @@ HasUnsupportedJoinWalker(Node *node, void *context)
 }
 
 
+/* QueryReferencesView returns true if the query references a view RTE */
+static bool
+QueryReferencesView(Query *subquery)
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: This isn't really an accurate name - there very well might not be a single reference to that view anymore.

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 10]
```diff
@@ -1560,7 +1630,9 @@ ExtractRangeTableRelationWalker(Node *node, List **rangeTableRelationList)
 	foreach(rangeTableCell, rangeTableList)
 	{
 		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
-		if (rangeTableEntry->rtekind == RTE_RELATION)
+		if (rangeTableEntry->rtekind == RTE_RELATION &&
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: Maybe it'd be better to negatively filter views here, rather than list all the types of tables we accept?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 11]
```diff
@@ -627,10 +627,25 @@ ErrorIfInsertSelectQueryNotSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		if (rangeTableEntry->rtekind == RTE_RELATION &&
+			rangeTableEntry->relkind == RELKIND_VIEW)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail("Cannot modify views")));
+		}
+	}
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Why do we need to check this for all the tables? Can't we just do the check for the target RTE?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 12]
```diff
@@ -2250,7 +2307,7 @@ UpdateRelationNames(Node *node, RelationRestrictionContext *restrictionContext)
 
 	newRte = (RangeTblEntry *) node;
 
-	if (newRte->rtekind != RTE_RELATION)
+	if (newRte->rtekind != RTE_RELATION || newRte->relkind != RELKIND_RELATION)
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Doesn't this break updating FDWs?

File: src/test/regress/expected/multi_view.out [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+	(SELECT l_suppkey, l_shipdate, count(*)
+		FROM air_shipped_lineitems GROUP BY l_suppkey, l_shipdate) supps
+	GROUP BY l_suppkey ORDER BY 2 DESC, 1 LIMIT 5;
+ERROR:  cannot perform distributed planning on this query
+DETAIL:  Subqueries with views are not supported yet
+-- logically same query without a view works fine
+SELECT l_suppkey, count(*) FROM
+	(SELECT l_suppkey, l_shipdate, count(*)
+		FROM lineitem_view_test WHERE l_shipmode = 'AIR' GROUP BY l_suppkey, l_shipdate) supps
+	GROUP BY l_suppkey ORDER BY 2 DESC, 1 LIMIT 5;
```
Comment (mtuncer) [inline src/test/regress/expected/multi_view.out]: This is a repartition query, so failure is expected in pushdown mode.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 11 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: targetRte (insertRte) points to an actual relation.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 14]
```diff
@@ -1531,12 +1531,14 @@ NewTableId(Index originalTableId, List *rangeTableList)
...
-			break;
+			return newTableId;
 		}
 
 		rangeTableIndex++;
 	}
 
+	ereport(ERROR, (errmsg("Unrecognized range table id %d", (int) originalTableId)));
+
 	return newTableId;
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_physical_planner.c]: Wouldn't it be better to remove the variable and return directly from the loop? Because this return is dead code...

File: src/backend/distributed/planner/multi_router_planner.c (hunk 11 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I don't understand your explanation?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 15]
```diff
@@ -832,11 +851,21 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 				break;
 			}
 
+			cteListContext = list_make1(query->cteList);
+			originalRelationId = FindOriginalRelationId(subqeryTargetEntry->expr,
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I'm a bit confused why we now need this - especially the CTE bits seem entirely unrelated to view support?

File: src/backend/distributed/planner/multi_router_planner.c (hunk 11 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: target RTE is determined from Query->resultRelation field. When we are inserting into a view, postgresql uses a pseudo-range table entry that points to the actual relation that would be receiving insert.

Therefore,  checking target RTE to see if it is a view does not provide any useful information. However, range table list contains 2 range table entries of type ```RELKIND_VIEW``` .  They are located at the 1st and 2nd places in range table list.

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 14 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_physical_planner.c]: ✅

File: src/backend/distributed/planner/multi_router_planner.c (hunk 15 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: This code portion checks if the Var corresponding to the partition column on the target table is actually a partition column in the source table.

When we are selecting from a view, TargetEntry->resorigtbl points to a view, and partition method or partition column calls fails due to that. I needed to do a recursive search based on target entry Var->attno to determine which relation it is coming from to perform partition column checks.

CTE involvement is unintended side effect here. If query contains a CTE we need to be able to traverse the CTE to be able to reach the source relation.

File: src/test/regress/sql/multi_create_table.sql [hunk 16]
```diff
@@ -136,3 +136,42 @@ SELECT repmodel FROM pg_dist_partition WHERE logicalrelid='mx_table_test'::regcl
...
+	o_custkey integer,
+	o_orderstatus char(1),
+	o_totalprice decimal(15,2),
+	o_orderdate date,
+	o_orderpriority char(15),
+	o_clerk char(15),
+	o_shippriority integer,
+	o_comment varchar(79) );
+
+SELECT create_distributed_table('orders_view_test', 'o_orderkey');
```
Comment (anarazel) [inline src/test/regress/sql/multi_create_table.sql]: We do we need a separate "_view_test" version of these tables? It seem much better to have a *single* _hash variant of lineitem that can be shared by multiple tests.

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 17]
```diff
@@ -1021,6 +1022,45 @@ TableEntryList(List *rangeTableList)
...
+	foreach(joinTreeTableIndexCell, joinTreeTableIndexList)
+	{
+		/*
+		 * Join tree's range table index starts from 1 in the query tree. But,
+		 * list indexes start from 0.
+		 */
+		int joinTreeTableIndex = lfirst_int(joinTreeTableIndexCell);
+		int rangeTableListIndex = joinTreeTableIndex - 1;
+		RangeTblEntry *rangeTableEntry =
+			(RangeTblEntry *) list_nth(rangeTableList, rangeTableListIndex);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_planner.c]: How about using rt_fetch() instead? Then you don't need to comment on the -1.

Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_planner.c]: makes sense

File: src/test/regress/sql/multi_create_table.sql (hunk 16 above)
Comment (mtuncer) [inline src/test/regress/sql/multi_create_table.sql]: that was to isolate this test from others. Also there are 3 more test cases that create ```lineitem_hash``` table with different settings. I guess we should stick to this usage until we implement populating a distributed table from a local table.  

We could load all data to a local table from file system once, and keep using that table as data source for other distributed tables used in tests.

**Comment by anarazel** [comment]
On 2017-01-04 02:49:31 -0800, Murat Tuncer wrote:
> mtuncer commented on this pull request.
> 
> 
> 
> > +	l_comment varchar(44) not null );
> +
> +SELECT create_distributed_table('lineitem_view_test', 'l_orderkey');
> +
> +CREATE TABLE orders_view_test (
> +	o_orderkey integer,
> +	o_custkey integer,
> +	o_orderstatus char(1),
> +	o_totalprice decimal(15,2),
> +	o_orderdate date,
> +	o_orderpriority char(15),
> +	o_clerk char(15),
> +	o_shippriority integer,
> +	o_comment varchar(79) );
> +
> +SELECT create_distributed_table('orders_view_test', 'o_orderkey');
> 
> that was to isolate this test from others. Also there are 3 more test
> cases that create ```lineitem_hash``` table with different settings. I
> guess we should stick to this usage until we implement populating a
> distributed table from a local table.

I think this is a ridiculously bad practice, which we shouldn't
continue.  I'm ok with naming it _hash_part or such for now, but we
should *NOT* continually create new tables for new tests. And by naming
it _view_test you're guaranteeing that it won't be reused.

**Comment by mtuncer** [comment]
_hash_part it is

**Comment by anarazel** [comment]
> _hash_part it is

Please open an issue for unifying some of the remaining table creations.

File: src/test/regress/sql/multi_create_table.sql [hunk 18]
```diff
@@ -136,3 +136,42 @@ SELECT repmodel FROM pg_dist_partition WHERE logicalrelid='mx_table_test'::regcl
 DROP TABLE mx_table_test; 
 
 SET citus.shard_replication_factor TO default;
+
+-- tables for view tests
```
Comment (anarazel) [inline src/test/regress/sql/multi_create_table.sql]: don't thinkt hat comment should be there.

File: src/test/regress/sql/multi_view.sql [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+-- This file contains test cases for view support. It verifies various
+-- Citus features: simple selects, aggregates, joins, outer joins
+-- router queries, single row inserts, multi row inserts via insert
+-- into select, multi row insert via copy commands.
+
+-- These are the settings that lineitem_hash_part and orders_hash_part tables
+-- are created with.
+SET citus.shard_count to 4;
+SET citus.shard_replication_factor to 1;
```
Comment (anarazel) [inline src/test/regress/sql/multi_view.sql]: Why's that relevant here?

File: src/test/regress/sql/multi_view.sql [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+
+CREATE VIEW air_shipped_lineitems AS SELECT * FROM lineitem_hash_part WHERE l_shipmode = 'AIR';
+
+-- join between view and table
+SELECT count(*) FROM orders_hash_part join air_shipped_lineitems ON (o_orderkey = l_orderkey);
+
+-- join between views
+SELECT count(*) FROM priority_orders join air_shipped_lineitems ON (o_orderkey = l_orderkey);
+
+-- count distinct on partition column is not supported, must be supported
```
Comment (anarazel) [inline src/test/regress/sql/multi_view.sql]: the phrasing here is unclear.

**Comment by mtuncer** [comment]
opened #1093

File: src/test/regress/sql/multi_view.sql [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+-- this will also fail due to same reason
+SELECT * FROM  lineitems_by_orderkey;
+
+-- however it would work if it is made router plannable
+SELECT * FROM  lineitems_by_orderkey WHERE l_orderkey = 100;
+
+DROP TABLE temp_lineitem CASCADE;
+DROP TABLE lineitem_hash_part CASCADE;
+DROP TABLE orders_hash_part CASCADE;
```
Comment (anarazel) [inline src/test/regress/sql/multi_view.sql]: it doesn't make sense to drop these here.

File: src/test/regress/sql/multi_create_table.sql (hunk 18 above)
Comment (mtuncer) [inline src/test/regress/sql/multi_create_table.sql]: right

File: src/test/regress/sql/multi_view.sql (hunk 19 above)
Comment (mtuncer) [inline src/test/regress/sql/multi_view.sql]: later in the test I create an empty distributed table and test insert into select. Tables must be colocated in order for that to work.

create_distributed_table did not have colocated_with support back then.
I will remove this and use colocate with.

File: src/test/regress/sql/multi_view.sql (hunk 20 above)
Comment (mtuncer) [inline src/test/regress/sql/multi_view.sql]: removing must be supported part

File: src/test/regress/sql/multi_view.sql (hunk 21 above)
Comment (mtuncer) [inline src/test/regress/sql/multi_view.sql]: ok

Comment (mtuncer) [inline src/test/regress/sql/multi_view.sql]: It made multi_metadata_sync fail. fixing it.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 22]
```diff
@@ -126,6 +126,8 @@ static void ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query,
 														   RangeTblEntry *subqueryRte,
 														   Oid *
 														   selectPartitionColumnTableId);
+Oid FindOriginalRelationId(Expr *columnExpression, List *parentQueryList,
+						   Query *query, bool *isPartitionColumn);
 static void AddUninstantiatedEqualityQual(Query *query, Var *targetPartitionColumnVar);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Huh? A non-static function prototype here?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 23]
```diff
@@ -3330,76 +3332,7 @@ bool
 IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: This function now shouldn't be named Recursive anymore, no?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 24]
```diff
@@ -50,6 +50,8 @@
 #include "utils/syscache.h"
 #include "utils/tqual.h"
 
+extern Oid FindOriginalRelationId(Expr *columnExpression, List *parentQueryList,
+								  Query *query, bool *isPartitionColumn);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Uh ;)

File: src/backend/distributed/planner/multi_router_planner.c [hunk 25]
```diff
@@ -871,6 +904,137 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 
 
 /*
+ * FindOriginalTableId recursively traverses query tree to find actual relation
+ * id that columnExpression refers to. It returns InvolidOid if Var is a non-relational
+ * or computed value from an inner subquery or cte. This is very similar to
+ * IsPartitionColumnRecursive.
+ */
+Oid
+FindOriginalRelationId(Expr *columnExpression, List *parentQueryList, Query *query,
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I think I'd prefer naming this ```FindReferencedTableColumn()``` (and returning both table / column) and moving the isPartitionColumn check to the caller (if required)?

It's also not true that this is similar to IsPartitionColumnRecursive() since the latter is now implemented using this function.

File: src/test/regress/expected/multi_insert_select.out [hunk 26]
```diff
@@ -1479,10 +1479,30 @@ COPY raw_events_first (user_id, value_1) FROM STDIN DELIMITER ',';
...
+SELECT count(*) FROM raw_events_second;
+ count 
+-------
+    13
+(1 row)
+
+-- inserting into views does not
+INSERT INTO test_view SELECT * FROM raw_events_second;
+ERROR:  cannot perform distributed planning for the given modification
+DETAIL:  Cannot modify views
```
Comment (anarazel) [inline src/test/regress/expected/multi_insert_select.out]: I think that message would be better if it were "cannot insert into view over distributed table" without any DETAIL.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 25 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: At least the parentQueryList parameter also needs to be documented.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 27]
```diff
@@ -871,6 +904,137 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
...
+
+		if (IsA(fieldExpression, Var))
+		{
+			candidateColumn = (Var *) fieldExpression;
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot push down this subquery"),
+							errdetail("Only references to column fields are supported")));
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Hm, that doesn't sound right in this place?

File: src/backend/distributed/planner/multi_router_planner.c (hunk 22 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: function definition and declaration is moved to multi_logical_optimizer.[c,h]

File: src/backend/distributed/planner/multi_router_planner.c (hunk 25 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: liked ```FindReferencedTableColumn()``` idea, adjusting the rest accordingly.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 27 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: This was from the original IsPartitionColumnRecursive function. However, none of the regression tests catch this part. Removing.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 23 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: not anymore

File: src/test/regress/expected/multi_insert_select.out (hunk 26 above)
Comment (mtuncer) [inline src/test/regress/expected/multi_insert_select.out]: done

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 24 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: taken care of

**Comment by metdos** [comment]
Hey @anarazel, @mtuncer is done with the changes :)

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 28]
```diff
@@ -3312,31 +3312,59 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
...
+ * that column is a partition column of the relation.
  *
- * Note that if the given expression is a field of a composite type, then this
- * function checks if this composite column is a partition column.
- *
- * Also, the function returns always false for reference tables given that reference
- * tables do not have partition column.
+ * Also, the function returns always false for reference tables given that
+ * reference tables do not have partition column.
  */
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Should we add a comment explaining that this doesn't support queries with CTEs?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 29]
```diff
@@ -3312,31 +3312,59 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
...
 bool
-IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
+IsPartitionColumn(Expr *columnExpression, Query *query)
 {
 	bool isPartitionColumn = false;
+	Oid relationId = InvalidOid;
+	Var *column = NULL;
+	FindReferencedTableColumn(columnExpression, NIL, query, &relationId, &column);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Ignorable nitpick: can you add a newline before FindReferencedTableColumn()?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 30]
```diff
@@ -3350,57 +3378,86 @@ IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
 		{
 			candidateColumn = (Var *) fieldExpression;
 		}
-		else
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Hm. Do we need a replacement check for this?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 31]
```diff
@@ -3350,57 +3378,86 @@ IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
...
 	{
 		Query *subquery = rangeTableEntry->subquery;
 		List *targetEntryList = subquery->targetList;
 		AttrNumber targetEntryIndex = candidateColumn->varattno - 1;
 		TargetEntry *subqueryTargetEntry = list_nth(targetEntryList, targetEntryIndex);
+		Expr *subColumnExpression = subqueryTargetEntry->expr;
 
-		Expr *subqueryExpression = subqueryTargetEntry->expr;
-		isPartitionColumn = IsPartitionColumnRecursive(subqueryExpression, subquery);
+		/* append current cteList to an existing cteListContext */
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: The list isn't called cteListContext anymore.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 32]
```diff
@@ -3350,57 +3378,86 @@ IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
...
+								  subquery, relationId, column);
 	}
 	else if (rangeTableEntry->rtekind == RTE_JOIN)
 	{
 		List *joinColumnList = rangeTableEntry->joinaliasvars;
 		AttrNumber joinColumnIndex = candidateColumn->varattno - 1;
 		Expr *joinColumn = list_nth(joinColumnList, joinColumnIndex);
 
-		isPartitionColumn = IsPartitionColumnRecursive(joinColumn, query);
+		/* cteListContext stays the same since still in the same query boundary */
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: same here.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 33]
```diff
@@ -1077,6 +1103,15 @@ ErrorIfModifyQueryNotSupported(Query *queryTree)
 			}
 
 			queryTableCount++;
+
+			/* we do not expect to see a view in modify query */
+			if (rangeTableEntry->relkind == RELKIND_VIEW)
+			{
+				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								errmsg("cannot perform distributed planning for the "
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: the errmsg is almost useless - the detail already seems like a better message. Something like "cannot modify views including distributed tables" would be better

Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: I thought I already fixed this, apparently not.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 32 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: done

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 31 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: done

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 30 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: I don't know a way to get to this state. There is no regression test hitting this one. 
If somehow  we get in this case  relationId would be InvalidOid, and column would be NULL. Callers would complain about not finding a partitionColumn instead of this error.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 28 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: added

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 29 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: ok

File: src/backend/distributed/planner/multi_router_planner.c (hunk 33 above)
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: Another note : it is impossible to hit this error anymore. 
```c
	Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
```
In the beginning of the function errors out before hitting this one.

**Comment by mtuncer** [comment]
@anarazel 2 issues remain open

1 - unreachable error message :  PartitionColumn() call in the beginning of the function is most likely an error check made by original author of the function.  We could live with this, but there is no test possible to hit this condition.

2 - not directly related commit regarding NewTableId() in physical planner.
We could commit this as a separate PR or a separate commit in this PR.

what do you think ?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 34]
```diff
@@ -3312,31 +3312,62 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
...
+}
+
+
+/*
+ * FindReferencedTableColumn recursively traverses query tree to find actual relation
+ * id, and column that columnExpression refers to. If columnExpression is a
+ * non-relational or computed/derived expression, the function returns InvolidOid for
+ * relationId and NULL for column. The caller should provide parent query list from
+ * top of the tree to this particular Query's parent. This argument is used to
+ * expand through CTEs that may be present in the query.
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: s/expand through/look into/?
